### PR TITLE
[Custom] Support build command with env variables

### LINKED ISSUE #3082: [Custom builder] Support build command with env variables
When the custom builder is configured with such command:
`docker build -t $IMAGES -f Dockerfile --ssh default .`,
it should run `sh -c "docker build -t $IMAGES -f Dockerfile --ssh default ."`

This allows using env variables.

### PR DESCRIPTION
**Pull request by dgageot**
Fixes #3082

### CODE REVIEW

File: pkg/skaffold/build/custom/custom.go [hunk 1]
```diff
@@ -68,9 +68,15 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
 
 func (b *ArtifactBuilder) retrieveCmd(out io.Writer, a *latest.Artifact, tag string) (*exec.Cmd, error) {
 	artifact := a.CustomArtifact
+
+	var cmd *exec.Cmd
 	split := strings.Split(artifact.BuildCommand, " ")
+	if len(split) > 1 {
+		cmd = exec.Command("sh", "-c", artifact.BuildCommand)
```
Comment (briandealwis) [inline pkg/skaffold/build/custom/custom.go]: This won't work on Windows.

File: pkg/skaffold/build/custom/custom.go [hunk 2]
```diff
@@ -68,9 +69,19 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
 
 func (b *ArtifactBuilder) retrieveCmd(out io.Writer, a *latest.Artifact, tag string) (*exec.Cmd, error) {
 	artifact := a.CustomArtifact
+
+	var cmd *exec.Cmd
 	split := strings.Split(artifact.BuildCommand, " ")
+	if len(split) > 1 {
```
Comment (nkubala) [inline pkg/skaffold/build/custom/custom.go]: if we're going to check the platform here anyway, then is there any point in checking `if len(split) > 1`? functionally `exec.Command(artifact.BuildCommand)` and `exec.Command("sh", "-c", artifact.BuildCommand)` should be the same thing right?

File: pkg/skaffold/build/custom/custom.go [hunk 3]
```diff
@@ -68,9 +69,19 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
...
+		if runtime.GOOS == "windows" {
```
Comment (nkubala) [inline pkg/skaffold/build/custom/custom.go]: also, it feels a little weird to have this check live right in the custom builder implementation. can this maybe be a function `GetShellCommand()` or something in `pkg/skaffold/util/cmd.go`?

File: pkg/skaffold/build/custom/custom.go (hunk 2 above)
Comment (dgageot) [reply pkg/skaffold/build/custom/custom.go]: It's not the same. I'll had a comment

File: pkg/skaffold/build/custom/custom.go (hunk 3 above)
Comment (dgageot) [reply pkg/skaffold/build/custom/custom.go]: I didn't want to make things too complicated. I'll what I can do.

**Comment by dgageot** [comment]
@briandealwis @nkubala Thanks for the review. I fixed the code for windows and added a comment that explains why we need to handle differently commands with or without args.

File: pkg/skaffold/build/custom/custom.go [hunk 4]
```diff
@@ -68,9 +69,26 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
...
+
 	split := strings.Split(artifact.BuildCommand, " ")
+	switch {
+	// When the user specifies a command without args, such as `./build.sh`,
+	// Skaffold can just exec this command.
+	case len(split) == 1:
+		cmd = exec.Command(artifact.BuildCommand)
+
+	// When the user specifies a command with args, such as `./build.sh -flag $IMAGE`,
+	// Skaffold has to shell out to sh in order for arguments to be evaluated.
```
Comment (briandealwis) [inline pkg/skaffold/build/custom/custom.go]: It's possible the executable is an environment variable too.  I define `$JAVA7`, `$JAVA8`, and `$JAVA11` in my environment to invoke the corresponding `java` binary.  Maybe we should just use `sh`/`cmd` always?

Comment (dgageot) [reply pkg/skaffold/build/custom/custom.go]: we could do that

File: pkg/skaffold/build/custom/custom.go [hunk 5]
```diff
@@ -68,9 +69,26 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
...
+	case runtime.GOOS != "windows":
```
Comment (balopat) [inline pkg/skaffold/build/custom/custom.go]: final nit: 

can we swap this? 
i.e. `default` would be `sh` and `runtime.GOOS == "windows"` would be `cmd.exe`

File: pkg/skaffold/build/custom/custom.go (hunk 4 above)
Comment (balopat) [reply pkg/skaffold/build/custom/custom.go]: I'm up for that. This split based determination is a "poor man's parsing" anyway - a proxy for "does this command have env vars"

**Pull request merged**